File: android/Cookpit/app/src/main/kotlin/com/github/kittinunf/cookpit/search/SearchDataController.kt
package com.github.kittinunf.cookpit.search

import com.github.kittinunf.cookpit.SearchController
import com.github.kittinunf.cookpit.SearchControllerObserver
import com.github.kittinunf.cookpit.SearchViewData
import com.github.kittinunf.reactiveandroid.MutableProperty

class SearchDataController : SearchControllerObserver() {

    private val controller = SearchController.create()

    private val _viewData = MutableProperty<SearchViewData>()

    val viewData by lazy { _viewData.observable }

    private val _recentItems = MutableProperty<List<String>>()

    val recentItems by lazy { _recentItems.observable }

    private val _loadings = MutableProperty(false)

    val loadings by lazy { _loadings.observable }

    var currentPage = 1

    init {
        controller.subscribe(this)
    }

    fun searchWith(key: String) {
        controller.reset()
        searchWithKey(key, 1)
    }

    private fun searchWithKey(key: String, page: Int) {
        currentPage = page
        controller.search(key, page.toByte())
    }

    fun searchNextPage(key: String) {
        searchWithKey(key, currentPage + 1)
    }

    fun fetchRecents() {
        _recentItems.value = controller.fetchRecents()
    }

    fun unsubscribe() {
        controller.unsubscribe()
    }

    override fun onBeginUpdate() {
        _loadings.value = true
    }

    override fun onUpdate(viewData: SearchViewData?) {
        _viewData.value = viewData
    }

    override fun onEndUpdate() {
        _loadings.value = false
    }

}

File: ios/Cookpit/Cookpit/SearchDataController.swift
//
//  SearchDataController.swift
//  Cookpit
//
//  Created by Kittinun Vantasin on 7/7/16.
//  Copyright © 2016 Cookpit. All rights reserved.
//

import Foundation
import RxSwift

class SearchDataController {

  private let controller = CPSearchController.create()!
  
  fileprivate let _viewData = Variable<CPSearchViewData?>(nil)
  
  lazy var viewData: Observable<CPSearchViewData> = {
    self._viewData.asObservable()
      .filter { $0 != nil }
      .map { $0! as CPSearchViewData }
      .distinctUntilChanged()
      .observeOn(MainScheduler.instance)
  }()
  
  private let _recentItems = Variable([String]())
  
  lazy var recentItems: Observable<[String]> = {
    self._recentItems.asObservable()
      .observeOn(MainScheduler.instance)
  }()
  
  fileprivate let _loadings = Variable(false)
  
  lazy var loadings: Observable<Bool> = {
    self._loadings.asObservable()
      .observeOn(MainScheduler.instance)
  }()
  
  lazy var errors: Observable<String> = {
    self._viewData.asObservable()
      .filter { $0 != nil }
      .map { $0! as CPSearchViewData }
      .filter { $0.error }
      .distinctUntilChanged()
      .map { $0.message }
      .observeOn(MainScheduler.instance)
  }()
  
  init() {
    controller.subscribe(self)
  }
  
  func unsubscribe() {
    controller.unsubscribe()
  }
  
  func fetchRecents() {
    let items = controller.fetchRecents()
    _recentItems.value = items
  }
  
  func searchWith(key: String) {
    controller.reset()
    searchWith(key: key, page: 1)
  }
  
  private func searchWith(key: String, page: Int) {
    self.controller.search(key, page: Int8(page))
  }
 
  deinit {
    controller.unsubscribe()
  }

}

extension SearchDataController : CPSearchControllerObserver {

  func onBeginUpdate() {
      _loadings.value = true
  }

  func onUpdate(_ viewData: CPSearchViewData) {
      _viewData.value = viewData
  }

  func onEndUpdate() {
      _loadings.value = false
  }

}
